### PR TITLE
Promote admission webhook e2e tests to conformance

### LINKED ISSUE #80767: Conformance Tests for AdmissionWebhook
- [x] write e2e tests for all admission operations
   - [x] https://github.com/kubernetes/kubernetes/pull/81381 (for resource APIs)
   - [ ] https://github.com/kubernetes/kubernetes/pull/82019 (for discovery docs)
   - [apisnoop beta admissionregistration coverage](https://apisnoop.cncf.io/?showUntested=true&showTested=true&showConformanceTested=true&zoomed=category-beta-admissionregistration)
- [x] switch to v1 admission APIs in e2e - https://github.com/kubernetes/kubernetes/pull/81558
    - [apisnoop stable admissionregistration coverage](https://apisnoop.cncf.io/?showUntested=true&showTested=true&showConformanceTested=true&zoomed=category-stable-admissionregistration)
- [x] remove skips/time.Sleeps - https://github.com/kubernetes/kubernetes/pull/81684
- [ ] [monitor for flakes](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=AdmissionWebhook&show-stale-tests=&width=5&sort-by-flakiness=) and promote to conformance tests - https://github.com/kubernetes/kubernetes/pull/81857

AdmissionWebhook should have tests for all stable operations.
Those tests should go through the conformance process before being released.
This ticket is to help track what operations are yet to be covered by conformance tests.

## 15 Untested Stable Operations need Conformance Tests

These stable endpoints should be hit by e2e tests tagged with [conformance].
There are tests currently hitting 6 of the corresponding beta endpoints that could be updated to use stable.
Tests should be written to exercise the remaining 9 untested stable endpoints as well.

## 12 Tests that hit 6 out of 15 Beta Operations

These twelve tests could be updated to use the stable API, and subsequently promoted to conformance.

[apisnoop.cncf.io tests for listAdmissionregistrationV1beta1ValidatingWebhookConfiguration](https://apisnoop.cncf.io/?showUntested=true&showTested=true&showConformanceTested=true&zoomed=operationId-beta-admissionregistration-listAdmissionregistrationV1beta1ValidatingWebhookConfiguration) / [test-grid](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=AdmissionWebhook&show-stale-tests=)

image

image

## Remaining 9 Untested Beta Operations

These will need tests written for them.

[apisnoop.cncf.io untested beta/admissionregistration](https://apisnoop.cncf.io/?showUntested=true&showTested=false&showConformanceTested=false&zoomed=category-beta-admissionregistration
) 

image

### PR DESCRIPTION
**Pull request by jpbetz**
Test stability status: https://k8s-testgrid.appspot.com/sig-release-master-blocking#gce-cos-master-default&width=5&sort-by-flakiness=&include-filter-by-regex=.*AdmissionWebhook.*

**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Admission webhoooks graduate to GA in 1.16. Conformance tests are now required for graduation to GA.

**Which issue(s) this PR fixes**:
Fixes #80767

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

- [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md)

/area conformance
@kubernetes/sig-architecture-pr-reviews @kubernetes/sig-api-machinery-pr-reviews @kubernetes/cncf-conformance-wg

/cc @liggitt @roycaihw @sttts @caesarxuchao 
/priority important-soon

### CODE REVIEW

File: test/e2e/apimachinery/webhook.go
Comment (johnbelamaric) [inline test/e2e/apimachinery/webhook.go]: Use RFC2119-style language. See https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-comment-metadata

for example:

Register an admission webhook configuration that denies configmap and pod creation. Attempt a creation of each; both MUST be denied.

This comment applies to all the descriptions.

File: test/e2e/apimachinery/webhook.go
Comment (johnbelamaric) [inline test/e2e/apimachinery/webhook.go]: nit: lowercase "should" (applies to all tests here)

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
I need to dig into the actual test functionality, but here's a quick couple comments that need addressing.

/assign

**Comment by BenTheElder** [comment]
FYI DNS failures like [these](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/81857/pull-kubernetes-conformance-kind-ipv6/1165029105624682496/) are currently expected on IPV6, pending a fix, and this job is NOT merge blocking currently. cc @aojea

File: test/e2e/apimachinery/webhook.go
Comment (jpbetz) [reply test/e2e/apimachinery/webhook.go]: All descriptions updated. I'll do this for the CRD Conformance PR as well.

File: test/e2e/apimachinery/webhook.go
Comment (jpbetz) [reply test/e2e/apimachinery/webhook.go]: All test names updated. I'll do this for the CRD Conformance PR as well.

**Comment by jpbetz** [comment]
Thanks for the quick first pass on syntactical issues @johnbelamaric. Feedback is applied.

**Comment by BenTheElder** [comment]
[hmm](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/81857/pull-kubernetes-e2e-kind/1165123611556581376) 🤔 
/retest

File: test/e2e/apimachinery/webhook.go
Comment (liggitt) [inline test/e2e/apimachinery/webhook.go]: lowercase should?

**Review by johnbelamaric — COMMENTED** [review]
This looks pretty good. There are some nits and some questions.

Also, is there a better way (or a plan for a better way) to check if a webhook config is in effect? The "marker" method is not very friendly.

File: test/e2e/apimachinery/webhook.go
Comment (aojea) [inline test/e2e/apimachinery/webhook.go]: this is failing with IPv6, I trill try to check tomorrow why

File: test/e2e/apimachinery/webhook.go
Comment (aojea) [inline test/e2e/apimachinery/webhook.go]: this is failing with IPv6 too, will try to check why

File: test/e2e/apimachinery/webhook.go
Comment (aojea) [reply test/e2e/apimachinery/webhook.go]: it's passing now, we'll try to monitor it 🤔

File: test/e2e/apimachinery/webhook.go
Comment (aojea) [reply test/e2e/apimachinery/webhook.go]: it's passing now, we'll try to monitor it 🤔

**Comment by roycaihw** [comment]
The ipv6 test is green (thanks @aojea!). We split out the test functionality change into a separate PR https://github.com/kubernetes/kubernetes/pull/82057, so we can keep the conformance promotion clean.

@johnbelamaric If things look good, I can squash and rebase on top of https://github.com/kubernetes/kubernetes/pull/82057.

**Comment by johnbelamaric** [comment]
yes, I think this is good too. my only concern is whether we need clusterAdmin for this. is that the case?

**Comment by liggitt** [comment]
Hoisting the answer from https://github.com/kubernetes/kubernetes/pull/81857#discussion_r318224073

Setup requires:
* read/write permissions to Deployment/Service/Secret objects, and read access to Endpoints (all of which are included in the default `edit` cluster role) to run the webhook pods and set up the relevant service
* read/write permissions on `{Mutating,Validating}WebhookConfiguration` objects (not included in any default role other than `cluster-admin`)
* read/write access to `CustomResourceDefinition` objects to declare a test type (not included in any default role other than `cluster-admin`)

Testing of the webhooks' efficacy requires:
* read/write permissions on pods/configmaps in the test namespaces (included in the default `edit` cluster role)
* read/write permissions on the test custom resource (not included in any default role other than `cluster-admin`)

> my only concern is whether we need clusterAdmin for this. is that the case?

admission configuration and custom resource definition are cluster-admin-level operations, so it is expected that testing those things requires that level of permission

**Comment by johnbelamaric** [comment]
Ok. That specifically violates the conformance rules. One other thought -
are these tests disruptive if run in a production cluster? The webhooks
installed apply to all namespaces, IIRC.

This is looking like an instance where the rules may be too restrictive. I
wonder if we need a separate set of tests that can only be run in a test
cluster.

**Comment by liggitt** [comment]
> That specifically violates the conformance rules.

I assume you mean this rule:
> it is non-privileged (e.g., does not require root on nodes, access to raw network interfaces, or cluster admin permissions)

I'm a bit at a loss on this aspect for tests verifying functionality of cluster-admin-targeted APIs... when combined with the requirement that GA APIs have conformance tests, that seems to imply there can be no GA APIs intended for use by cluster admins.

> One other thought - are these tests disruptive if run in a production cluster? The webhooks
installed apply to all namespaces, IIRC.

No, they are scoped to namespaces or objects with an e2e-specific label exactly to avoid being disruptive. See the checks here:

https://github.com/kubernetes/kubernetes/blob/06dc8cf4cb0d3bdc7b718392a66ca82678d9982f/test/e2e/apimachinery/webhook.go#L2053-L2081

**Comment by johnbelamaric** [comment]
Yes, it’s not really entirely tenable as written.

**Comment by johnbelamaric** [comment]
Let me dig into the motivation for that rule and get back to you ASAP.

**Comment by johnbelamaric** [comment]
Ok, we need to revise the language in the criteria - there are already existing tests that require various levels of privilege. So we don’t want to strictly block on this. It is more around avoiding _unnecessary_ privileges. We need to provide the ability to segregate the tests based on privileges needed. 

Along those lines, it makes sense to tag these tests with a [Privileged:ClusterAdmin] tag (which is a new tag).

This is getting a little ahead of the policy but we can always tweak the tag later if the final policy is a bit different.

**Comment by roycaihw** [comment]
thanks @liggitt @johnbelamaric. I'm adding the [Privileged:ClusterAdmin] tag

**Comment by roycaihw** [comment]
rebased and promoted the discovery doc test https://github.com/kubernetes/kubernetes/pull/82019, to cover all endpoints for conformance

**Comment by smarterclayton** [comment]
Originally conformance was "make sure apps run on kube clusters the same".  We limited privilege because lots of clusters won't allow those privileges.  We want conformance to apply to more than just admin function.  We really, really, really need either profiles or to drop the requirements for privilege in the short term.

**Comment by johnbelamaric** [comment]
/lgtm

Note for conformance approvers: These tests require clusterAdmin, because this functionality is inherently privileged. To mitigate this in environments where the tests cannot be run as clusterAdmin, we added a [Privileged:ClusterAdmin] tag to that they can be easily excluded. This is a new tag and I will propose an update to the docs to cover this case.

**Comment by smarterclayton** [comment]
/approve

This is such a fundamental part of Kube that I think we should get it in conformance, then use this as a focus to sort out profiles.  You can be conformant for apps without allowing webhook registration, but you can't be conformant for extension without it.

**Pull request merged**
